gpio/src/gpio.rs

```diff
@@ -0,0 +1,943 @@
+///
+/// This trait is introduced for development purposes only, and should not
+/// be used outside of this crate. The purpose of this trait is to provide a
+/// mock implementation for the GPIO driver so that we can test the GPIO
+/// functionality without the need of a physical device.
+pub(crate) trait GpioDevice: Send + Sync + 'static {
+    fn open(device: u32) -> Result<Self>
+    where
+        Self: Sized;
+
+    fn get_num_lines(&self) -> Result<u16>;
+    fn get_line_name(&self, gpio: u16) -> Result<String>;
+    fn get_direction(&self, gpio: u16) -> Result<u8>;
+    fn set_direction(&self, gpio: u16, dir: u8, value: u32) -> Result<()>;
+    fn get_value(&self, gpio: u16) -> Result<u8>;
+    fn set_value(&self, gpio: u16, value: u32) -> Result<()>;
+
+    fn set_irq_type(&self, gpio: u16, value: u16) -> Result<()>;
+    fn wait_for_interrupt(&self, gpio: u16) -> Result<()>;
+}
+
+pub(crate) struct PhysDeviceState {
+    // See wait_for_interrupt() for explanation of Arc.
+    request: Option<Arc<LineRequest>>,
+    buffer: Option<EdgeEventBuffer>,
+}
+
+pub(crate) struct PhysDevice {
+    chip: Chip,
+    ngpio: u16,
+    state: Vec<RwLock<PhysDeviceState>>,
+}
+
+unsafe impl Send for PhysDevice {}
+unsafe impl Sync for PhysDevice {}
+
+impl GpioDevice for PhysDevice {
+    fn open(device: u32) -> Result<Self>
+    where
+        Self: Sized,
+    {
+        let path = format!("/dev/gpiochip{}", device);
+        let chip = Chip::open(&path).map_err(Error::GpiodFailed)?;
+        let ngpio = chip.get_num_lines() as u16;
+
+        // Can't set a vector to all None easily
+        let mut state: Vec<RwLock<PhysDeviceState>> = Vec::new();
+        state.resize_with(ngpio as usize, || {
+            RwLock::new(PhysDeviceState {
+                request: None,
+                buffer: None,
+            })
+        });
+
+        Ok(PhysDevice { chip, ngpio, state })
+    }
+
+    fn get_num_lines(&self) -> Result<u16> {
+        Ok(self.ngpio)
+    }
+
+    fn get_line_name(&self, gpio: u16) -> Result<String> {
+        let line_info = self
+            .chip
+            .line_info(gpio.into())
+            .map_err(Error::GpiodFailed)?;
+
+        Ok(line_info.get_name().unwrap_or("").to_string())
+    }
+
+    fn get_direction(&self, gpio: u16) -> Result<u8> {
+        let line_info = self
+            .chip
+            .line_info(gpio.into())
+            .map_err(Error::GpiodFailed)?;
+
+        Ok(
+            match line_info.get_direction().map_err(Error::GpiodFailed)? {
+                Direction::AsIs => VIRTIO_GPIO_DIRECTION_NONE,
+                Direction::Input => VIRTIO_GPIO_DIRECTION_IN,
+                Direction::Output => VIRTIO_GPIO_DIRECTION_OUT,
+            },
+        )
+    }
+
+    fn set_direction(&self, gpio: u16, dir: u8, value: u32) -> Result<()> {
+        let mut config = LineConfig::new().map_err(Error::GpiodFailed)?;
+        let state = &mut self.state[gpio as usize].write().unwrap();
+
```

Comment:
Shouldn't this function just be concerned with setting the direction of the line rather than both the line and value?  This seems at odd with get_direction() above and get/set_value() below.

Reply:
Setting of a line with `OUT` direction always comes with a value to be set for the line, else it can potentially break the hardware in the worst case.

If the direction is out, we need to configure value as well. And we can't set a value before configuring the direction. This is how this is done generally, everywhere, including in kernel drivers.

Reply:
Ok

Reply:
You could just rename the function to be explicit: set_direction_and_state()

Reply:
> You could just rename the function to be explicit: set_direction_and_state()

This is a trait function, I would be required to change it for each implementation then. And in that case it would rather be better to have two routines, set_direction_out_with_value() and set_direction_in().

So I thought maybe set_direction() is just fine.